### DISCUSSION TITLE
Build picolib.a

### DISCUSSION THREAD

**Discussion by PontusO** [earlephilhower/arduino-pico#286] [Q&A] [answered]
Hi guys. Is there any documentation available on how to build picolib.a ? I'd like to experiment with different sleep modes from the sdk and those pico-extras files does not seem to have been included in the archive.

**Comment by earlephilhower** [comment, marked as answer]
It's described in https://github.com/earlephilhower/arduino-pico/tree/master/tools .  Just cd into the tools/libpico directory and `./make-libpico.sh` to build.

**Reply by bzisl** [reply]
I'm working on macOS, and I tried, but it fails. I will try at Windows also.
Do you need something more installed?

**Reply by earlephilhower** [reply]
You need to be able to build a Pico SDK app using their tools.  See the PIO SDK docs for what you need to install.  They have Win, Mac, and Linux sections IIRC.

**Comment by PontusO** [comment]
Excellent thanks. Do you have any plans on including the sleep and dormant modes in the rp2040 class ?

**Reply by earlephilhower** [reply]
If you want to add the add'l SDK sources to `libpico.a`, just send in a PR. Should just be 1-2 lines in the Cmakefile.txt list.

I've not really looked at the sleep modes, but it seems like something that could be added to the RP2040 class simply enough as long as the main SRAMs stay powered.  There might be concerns about the peripherals (i.e. I'd bet the PWM clock, for example, would be disabled and need to be manually re-enabled if there were analogWrites active).

There are also some parts of the core where it would be possible to disable or selectively enable clocks to peripherals, but as of now that's not done.  For example, on the 1st write, all PWM generator clocks are turned on even though you could be smart about it and only enable the one being written to.  PIO emulated peripherals are also the same way.  Since it's a 1-man show here, I preferred expediency and correctness over most-optimized. :)
